### PR TITLE
feat: Introduce ConversationFolderSelectionUseCase - WPB-12010

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-12010" title="WPB-12010" target="_blank">TaskWPB-12010</a>  [iOS] Folder Managements View Models 
  </td></table>
  <br />
 

---

In preparation for re-writing the Move to Folder feature in SwiftUI with this PR we introduce one of the use cases we're going to use: `ConversationFolderSelectionUseCase` More specifically:

1. New `ToFolderMovableConversation` Protocol
  - Dedicated protocol for folder movement capabilities
  - Single responsibility: `moveToFolder(_ folder: LabelType)`

2. Use Case Pattern Implementation
  - `ConversationFolderSelectionUseCase` for handling folder selection

3. Testing
  - `MockLabelType` for simulating folders
  - ConversationFolderSelectionUseCaseTests

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### CODE REVIEW

File: wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireDataModel
+
+// MARK: - ConversationFolderSelectionUseCaseProtocol
+
+protocol ConversationFolderSelectionUseCaseProtocol {
```
Comment (caldrian) [inline wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift]:
```suggestion
public protocol ConversationFolderSelectionUseCaseProtocol {
```

Shall we move it to a dedicated file?

File: wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireDataModel
```
Comment (caldrian) [inline wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift:19]: Why is this use case not in WireDataModel itself?

**Review by caldrian — APPROVED** [review]

File: wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift:19]: We usually keep the useCases in SE.

File: wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+// MARK: - ConversationFolderSelectionUseCaseProtocol
+
+public protocol ConversationFolderSelectionUseCaseProtocol {
```
Comment (samwyndham) [inline wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift:23]: I don't think this protocol is needed

Instead I expect you will need to define a protocol in your `WireUI` target and conform the protocol to it outside (i.e Wire-iOS target).

Personally I would avoid creating the protocol until you actually need it.

**Review by samwyndham — APPROVED** [review]

File: wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/ConversationFolderSelectionUseCase.swift:23]: Let's keep it for now and see later if we indeed need it or not. Unclear for me still.

**Pull request merged**